pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,500 @@
+		connWg:           &sync.WaitGroup{},
+		incomingMessages: make(chan *events.EventList),
+		nodes:            make(map[types.NodeID]*nodeInfo),
+		logger:           logger,
+		ownID:            ownID,
+		host:             h,
+	}, nil
+}
+
+func (t *Transport) Start() error {
+	t.logger.Log(logging.LevelDebug, "mir handler is starting", "ownID", t.ownID, "addrs", t.host.Addrs())
+	t.host.SetStreamHandler(ProtocolID, t.mirHandler)
+	return nil
+}
+
+func (t *Transport) Stop() {
+	t.logger.Log(logging.LevelDebug, "Stopping libp2p transport.", "ownID", t.ownID)
+	defer t.logger.Log(logging.LevelDebug, "Stopping libp2p transport finished.", "ownID", t.ownID)
+
+	t.nodesLock.Lock()
+	defer t.nodesLock.Unlock()
+
+	for id, info := range t.nodes {
+		if info.Stream == nil {
+			continue
+		}
+		t.logger.Log(logging.LevelDebug, "Closing connection", "to", id)
+
+		if err := info.Stream.Close(); err != nil {
+			t.logger.Log(logging.LevelError, "Could not close connection.", "nodeID", id, "err", err)
+			continue
+		}
+
+		t.logger.Log(logging.LevelDebug, "Closed connection", "to", id)
+	}
+
+	t.host.RemoveStreamHandler(ProtocolID)
+
+	if err := t.host.Close(); err != nil {
+		t.logger.Log(logging.LevelError, "Could not close libp2p", "ownID", t.ownID, "err", err)
+	} else {
+		t.logger.Log(logging.LevelDebug, "libp2p host closed", "ownID", t.ownID)
+	}
+
+	t.connWg.Wait()
+}
+
+func (t *Transport) CloseOldConnections(newNodes map[types.NodeID]types.NodeAddress) {
+	t.connWg.Add(1)
+
+	go func() {
+		defer t.connWg.Done()
+
+		for nodeID, nodeInfo := range t.nodes {
+			// Close an old connection to a node if we don't need to connect to it further.
+			if _, foundInNewNodes := newNodes[nodeID]; !foundInNewNodes {
+				t.logger.Log(logging.LevelDebug, "Closing old connection", "src", t.ownID, "dst", nodeID)
+
+				if nodeInfo.Stream != nil {
+					info, err := peer.AddrInfoFromP2pAddr(nodeInfo.Addr)
+					if err != nil {
+						t.logger.Log(logging.LevelError, "failed to parse addr", "src", t.ownID, "addr", nodeInfo.Addr, "err", err)
+						continue
+					}
+					err = t.host.Network().ClosePeer(info.ID)
+					if err != nil {
+						t.logger.Log(logging.LevelError, "Could not close old connection to node", "src", t.ownID, "dst", nodeID, "err", err)
+						continue
+					}
+				} else {
+					t.logger.Log(logging.LevelDebug, "No stream for old connection", "src", t.ownID, "dst", nodeID)
+				}
+
+				t.nodesLock.Lock()
+				delete(t.nodes, nodeID)
+				t.nodesLock.Unlock()
+			}
+		}
+	}()
+}
+
+func (t *Transport) Connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
+	if len(nodes) == 0 {
+		t.logger.Log(logging.LevelWarn, "no nodes to connect to")
+		return
+	}
+
+	t.nodesLock.Lock()
+	for id, addr := range nodes {
+		_, found := t.nodes[id]
+		if !found {
+			t.nodes[id] = &nodeInfo{
+				ID:        id,
+				Addr:      addr,
+				IsOpening: false,
+				Stream:    nil,
+			}
+
+		}
+	}
+	t.nodesLock.Unlock()
+
+	t.connWg.Add(1)
```

Comment:
Why are you using this `WaitGroup` here, instead of including all this waiting logic directly in `connect` (It feels harder to reason about if the waiting logic is scattered through different functions)? Can the high-level function `Connect` be called by several routines?

Reply:
Agree, I moved this inside of `connect` function